gotrue/types.py

```diff
@@ -0,0 +1,346 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from enum import Enum
+from inspect import signature
+from json import dumps
+from time import time
+from typing import Any, Callable, Optional, Type, TypeVar
+
+T = TypeVar("T")
+
+
+def parse_none(v: Optional[T], f: Callable[[Any], T]) -> Optional[T]:
+    if v is None:
+        return None
+    return f(v)
+
+
+def parse_dict(cls: Type[T], **json: dict) -> T:
+    cls_fields = {field for field in signature(cls).parameters}
+    native_args, new_args = {}, {}
+    for name, val in json.items():
+        if name in cls_fields:
+            native_args[name] = val
+        else:
+            new_args[name] = val
+    ret = cls(**native_args)
+    for new_name, new_val in new_args.items():
+        setattr(ret, new_name, new_val)
+    return ret
+
+
+@dataclass
+class APIError(BaseException):
+    msg: str
+    code: int
+
+    def __post_init__(self) -> None:
+        self.msg = str(self.msg)
+        self.code = int(str(self.code))
+
+    @classmethod
+    def from_dict(cls, data: dict) -> APIError:
+        if "msg" in data and "code" in data:
+            return APIError(
+                msg=data["msg"],
+                code=data["code"],
+            )
+        if "error" in data and "error_description" in data:
+            try:
+                code = int(data["error"])
+            except ValueError:
+                code = -1
+            return APIError(
+                msg=data["error_description"],
+                code=code,
+            )
+        return parse_dict(cls, **data)
+
+    def to_dict(self) -> dict[str, Any]:
+        return {
+            "msg": self.msg,
+            "code": self.code,
+        }
+
+    def __str__(self) -> str:
+        return dumps(self.to_dict())
+
+    def __repr__(self) -> str:
+        return self.__str__()
+
+
+@dataclass
+class CookieOptions:
+    name: str
+    """The name of the cookie. Defaults to `sb:token`."""
+    lifetime: int
+    """The cookie lifetime (expiration) in seconds. Set to 8 hours by default."""
+    domain: str
+    """The cookie domain this should run on.
+    Leave it blank to restrict it to your domain."""
+    path: str
+    same_site: str
+    """SameSite configuration for the session cookie.
+    Defaults to 'lax', but can be changed to 'strict' or 'none'.
+    Set it to false if you want to disable the SameSite setting."""
+
+    def __post_init__(self) -> None:
```

Comment:
I also prefer using `pydantic` models (I love `pydantic` ❤️ )

In this issue https://github.com/supabase-community/gotrue-py/issues/15 I comment on the alternative of using `pydantic`, but they told me to use `dataclasses`.

What do you think @J0 ? `pydantic` is really fast because I use `C` internally.

Reply:
From @J0 comment in #50, I understand that with dataclasses, he meant the abstraction of having classes that managed data, and that it was up to you to choose what to use in the implementation (dataclasses, attrs, pydantic or others). Correct me if I'm wrong Joel 🙂

Reply:
Perfect, if they confirm that this is the case, I implement it with `pydantic`, it would make me happier 😄

Reply:
👍🏻 pydantic would be a good fit here

Reply:
In case you want to use `pydantic`, I already have the implementation done in another branch. https://github.com/leynier/gotrue-py/tree/feat/pydantic

Reply:
Awesome! Let's wait for @J0's opinion, maybe a day or two more, since he might have more time in the weekend.

If he doesn't answer in a couple days, since we three think it's a good idea to use pydantic and I rarely think he would find it a bad idea, then you could merge pydantic branch into this PR's branch. @leynier

Reply:
Hey all, sorry for the delays on this. Will revert by end of day.

Reply:
Sorry everyone for the delayed replies. I am also for pydantic and @dreinon is correct #15 is referring to  the abstraction of having classes that managed data and any sort of implementation is fine.